src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,772 @@
+#ifndef __STRINGS_HH
+#define __STRINGS_HH
+
+#include <array>
+#include <cstring>
+#include <cerrno>
+#include <limits>
+#include <type_traits>
+#include <unistd.h>
+
+#include "platform-compat.hh"
+#include "logger.hh"
+#include "helpers.hh"
+
+namespace xamarin::android::internal
+{
+	constexpr size_t SENSIBLE_TYPE_NAME_LENGTH = 128;
+	constexpr size_t SENSIBLE_PATH_MAX = 256;
+	// 64-bit unsigned or 64-bit signed with sign
+	constexpr size_t MAX_INTEGER_DIGIT_COUNT_BASE10 = 21;
+
+#if DEBUG
+	constexpr bool BoundsCheck = true;
+#else
+	constexpr bool BoundsCheck = false;
+#endif
+
+	class string_segment
+	{
+	public:
+		force_inline bool initialized () const noexcept
+		{
+			return !_fresh;
+		}
+
+		force_inline const char* start () const noexcept
+		{
+			return _start;
+		}
+
+		force_inline size_t length () const noexcept
+		{
+			return _length;
+		}
+
+		force_inline bool empty () const noexcept
+		{
+			return length () == 0;
+		}
+
+		force_inline bool equal (const char *s) const noexcept
+		{
+			if (s == nullptr)
+				return false;
+
+			return equal (s, strlen (s));
+		}
+
+		force_inline bool equal (const char *s, size_t s_length) const noexcept
+		{
+			if (!can_access (s_length)) {
+				return false;
+			}
+
+			if (length () != s_length) {
+				return false;
+			}
+
+			if (length () == 0) {
+				return true;
+			}
+
+			return memcmp (_start, s, length ()) == 0;
+		}
+
+		force_inline bool starts_with (const char *s) const noexcept
+		{
+			if (s == nullptr)
+				return false;
+
+			return starts_with (s, strlen (s));
+		}
+
+		force_inline bool starts_with (const char *s, size_t s_length) const noexcept
+		{
+			if (s == nullptr || !can_access (s_length)) {
+				return false;
+			}
+
+			if (length () < s_length) {
+				return false;
+			}
+
+			return memcmp (start (), s, s_length) == 0;
+		}
+
+		template<size_t Size>
+		force_inline bool starts_with (const char (&s)[Size]) const noexcept
+		{
+			return starts_with (s, Size);
+		}
+
+		force_inline bool has_at (const char ch, size_t index) const noexcept
+		{
+			if (!can_access (index)) {
+				return false;
+			}
+
+			return start ()[index] == ch;
+		}
+
+		force_inline ssize_t find (const char ch, size_t start_index) const noexcept
+		{
+			if (!can_access (start_index)) {
+				return -1;
+			}
+
+			while (start_index <= length ()) {
+				if (start ()[start_index] == ch) {
+					return static_cast<ssize_t>(start_index);
+				}
+				start_index++;
+			}
+
+			return -1;
+		}
+
+		template<typename T>
+		force_inline bool to_integer (T &val, size_t start_index = 0, int base = 10) const noexcept
```

Comment:
Why do we want to inline a member function as large as this?

Reply:
It's not big at all. There's a lot of fluff in the code, syntactic niceties. And when this code is compiled, it is reduced to basically a single `strto*ll` call + some error checking, the same as if you wrote it by hand. The entire function takes 124 bytes of `aarch64` assembly:

```
 5c4:   350007e9        cbnz    w9, 6c0 <xamarin::android::internal::MonodroidRuntime::parse_runtime_args(xamarin::android::internal::dynamic_local_string<93ul, char>&, xamarin::android::internal::MonodroidRuntime::RuntimeOptions*)+0x6c0>
 5c8:   f1002518        subs    x24, x8, #0x9
 5cc:   540007a3        b.cc    6c0 <xamarin::android::internal::MonodroidRuntime::parse_runtime_args(xamarin::android::internal::dynamic_local_string<93ul, char>&, xamarin::android::internal::MonodroidRuntime::RuntimeOptions*)+0x6c0>  // b.lo, b.ul, b.last
 5d0:   f85383a9        ldur    x9, [x29, #-200]
 5d4:   b4000769        cbz     x9, 6c0 <xamarin::android::internal::MonodroidRuntime::parse_runtime_args(xamarin::android::internal::dynamic_local_string<93ul, char>&, xamarin::android::internal::MonodroidRuntime::RuntimeOptions*)+0x6c0>
                        char s[slen + 1];
 5d8:   91001d08        add     x8, x8, #0x7
 5dc:   910003ea        mov     x10, sp
 5e0:   927ced08        and     x8, x8, #0xfffffffffffffff0
 5e4:   cb080157        sub     x23, x10, x8
 5e8:   910002ff        mov     sp, x23
                        memcpy (s, start () + start_index, slen);
 5ec:   91002521        add     x1, x9, #0x9
 5f0:   aa1703e0        mov     x0, x23
 5f4:   aa1803e2        mov     x2, x24
 5f8:   94000000        bl      0 <memcpy>
                        s[slen] = '\0';
 5fc:   38386aff        strb    wzr, [x23, x24]
                        errno = 0;
 600:   94000000        bl      0 <__errno>
 604:   aa0003f8        mov     x24, x0
 608:   b900001f        str     wzr, [x0]
                                ret = strtoll (s, &endp, base);
 60c:   d102e3a1        sub     x1, x29, #0xb8
 610:   52800142        mov     w2, #0xa                        // #10
 614:   aa1703e0        mov     x0, x23
 618:   94000000        bl      0 <strtoll>
```